rpk: support for rpk shadow in Redpanda Cloud

r-vasquez opened the pull request:
This PR introduces the support for `rpk shadow` against Redpanda Cloud clusters, the supported commands are:

- create
- list        (now supports --format json/yaml flag)
- status  (now supports --format json/yaml flag)
- describe
- update
- delete
- failover
- config generate

### **Main differences between SH and Cloud**

**Profile setup**
A user **must be** logged in Redpanda Cloud (using `rpk cloud login`) and select the **shadow cluster** in order to manage the shadow links. A profile will be created when you select the cluster, as always.

If the user does not have a profile setup with the shadow cluster, the commands won't work properly.

**The configuration file:**
The configuration file has a new field `cloud_options` that allow users to specify the source and shadow redpanda_id. A user can either specify the source_redpanda_id, or a bootstrap server URL in `client_options.bootstrap_servers`.

These changes are reflected in the sample config one can get in `rpk shadow config generate --for-cloud`
```
cloud_options:
  source_redpanda_id: m7xtv2qq5njbhwruk88f
  resource_group_id: 3f72b1a9-5c4e-4d82-9ab7-1c5f8e9d3a6b
  shadow_redpanda_id: p9skc1dd3fmzgvquj66h
```

**Forced Deletion**
By default, deleting a shadow link in Redpanda cloud defaults to a Force delete, meaning that any active shadow topic will be failed over.


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Improvements

* rpk shadow is now supported in Redpanda Cloud

r-vasquez committed:
rpk: move cloud auth token to its own package

This will allow us to use this lightweight
package as helper in other packages w/o
circular dependencies with the bigger
package (oauth). We are expanding our usage
as we introduce more publicAPI commands.

An example usage is introduced in the
controlplane client where we are now able
to validate the token before creating the
client.

r-vasquez committed:
rpk: initial support for rpk shadow in cloud

This commit includes the addition of the
ShadowLink service to our Controlplane
client holder and includes initial support
for `create` and `list`.

Full detail:
- Include new configuration field 'cloud_options'
  to hold all Redpanda Cloud exclusive settings.
- Add new mappers to translate our config file
  into the Cloud proto.
- Support for `shadow create`.
- Support for `shadow list`, including the
  --format json/yaml flag.
- Updated help text.

c-julin commented on src/go/rpk/pkg/oauth/authtoken/authtoken.go:
```diff
@@ -0,0 +1,95 @@
+// Copyright 2024 Redpanda Data, Inc.
```
2025

c-julin commented on src/go/rpk/pkg/oauth/authtoken/authtoken.go:
```diff
@@ -0,0 +1,95 @@
+		if err == nil {
+			return false, nil
+		}
+		switch err.Error() {
```
single case as a switch doesn't make sense think it should be an if statement

r-vasquez replied:
I swear I've seen a linter complaining about that. I wonder why we haven't caught this in a long time (this is the same code, just moved over). I'll clean it 👍

c-julin commented on src/go/rpk/pkg/cli/shadow/config.go:
```diff
@@ -43,16 +43,20 @@ func newShadowConfigCommand(fs afero.Fs, p *config.Params) *cobra.Command {
+	var (
+		outputPath    string
+		printTemplate bool
+		forCloud      bool
```
i think this should just be --cloud , the for is implied

c-julin commented on src/go/rpk/pkg/cli/shadow/config.go:
```diff
@@ -196,6 +200,29 @@ func generateSampleConfig() *ShadowLinkConfig {
+			ResourceGroupID:       "3f72b1a9-5c4e-4d82-9ab7-1c5f8e9d3a6b",
+		}
+		// Replace Cloud bootstrap URL to make it look like a Cloud endpoint.
+		slcfg.ClientOptions.BootstrapServers = []string{"seed-332d9d9e.p9skc1dd3fmzgvquj66h.fmc.prd.cloud.redpanda.com:9092"}
```
in the sample send either source redpanda id or the bootstrap servers think both will confuse users

c-julin commented on src/go/rpk/pkg/cli/shadow/config.go:
```diff
@@ -196,6 +200,29 @@ func generateSampleConfig() *ShadowLinkConfig {
+		slcfg.ClientOptions.BootstrapServers = []string{"seed-332d9d9e.p9skc1dd3fmzgvquj66h.fmc.prd.cloud.redpanda.com:9092"}
+
+		// This is confusing on Cloud, user is already providing the Redpanda ID.
+		slcfg.ClientOptions.SourceClusterID = ""
```
does this omit if empty?

r-vasquez replied:
Yes, https://github.com/redpanda-data/redpanda/blob/dev/src/go/rpk/pkg/cli/shadow/types.go#L38

c-julin commented on src/go/rpk/pkg/cli/shadow/create.go:
```diff
@@ -77,12 +86,39 @@ Create a Shadow Link without confirmation prompt:
+				out.MaybeDieErr(err)
+				if isComplete {
+					out.Exit(successMsgTmpl, slCfg.Name, op.Msg.GetOperation().GetResourceId())
+				} else {
```
i dont think these else branches are necessary, both this and the one on 111 as you're exiting  at  all branches

c-julin commented on src/go/rpk/pkg/cli/shadow/create.go:
```diff
@@ -110,9 +146,18 @@ func printShadowLinkCfgOverview(slCfg *ShadowLinkConfig) {
+		if slCfg.CloudOptions.SourceRedpandaID != "" {
+			tw.Print("Source Redpanda ID:", slCfg.CloudOptions.SourceRedpandaID)
+		}
+		tw.Print("Resource Group ID:", slCfg.CloudOptions.ResourceGroupID)
```
i think we should still support bootstrap servers in cloud, if they want to shadow from a self hosted cluster or a confluent/msk cluster

r-vasquez replied:
They are supported though, this is just printing the overview, and the bootstrap server is part of the clientOptions. That's why it's below on L156

c-julin commented on src/go/rpk/pkg/cli/shadow/create.go:
```diff
@@ -123,7 +168,8 @@ func validateParsedShadowLinkConfig(slCfg *ShadowLinkConfig) error {
 	}
-	if len(slCfg.ClientOptions.BootstrapServers) == 0 {
+	// Cloud configuration does not require bootstrap servers.
+	if slCfg.CloudOptions == nil && len(slCfg.ClientOptions.BootstrapServers) == 0 {
```
if the error you're returning is about source server then the check should be `slCfg.CloudOptions.SourceRedpandaID == nil`

r-vasquez replied:
This check is for self_hosted only: if CloudOptions are nil and don't have bootstrap servers, then fail.

I will add a new one below L191 that checks that either source_redpanda_id OR clientOptions.bootstrapServers must be set

c-julin commented on src/go/rpk/pkg/cli/shadow/create.go:
```diff
@@ -144,5 +190,55 @@ func validateParsedShadowLinkConfig(slCfg *ShadowLinkConfig) error {
+		if slc.ResourceGroupID == "" {
+			return fmt.Errorf("resource_group_id is required in cloud options")
+		}
+		if slCfg.ClientOptions != nil && slCfg.ClientOptions.TLSSettings != nil && slCfg.ClientOptions.TLSSettings.TLSFileSettings != nil {
```
do we need to tell the user? typically if a user provided extra fields it would just not accept

r-vasquez replied:
It wouldn't hurt. Since we are reusing the same configuration file the mapper will silently ignore it. So I prefer a warning instead of doing something they might not be expecting silently

c-julin commented on src/go/rpk/pkg/cli/shadow/create.go:
```diff
@@ -144,5 +190,55 @@ func validateParsedShadowLinkConfig(slCfg *ShadowLinkConfig) error {
 }
+
+func waitForCloudShadowLink(ctx context.Context, cloudClient *publicapi.CloudClientSet, linkID string) (isCompleted bool, err error) {
+	maxRetries := 5
```
these can be consts so they're not assigned everytime on runtime?

c-julin commented on src/go/rpk/pkg/cli/shadow/create.go:
```diff
@@ -144,5 +190,55 @@ func validateParsedShadowLinkConfig(slCfg *ShadowLinkConfig) error {
+func waitForCloudShadowLink(ctx context.Context, cloudClient *publicapi.CloudClientSet, linkID string) (isCompleted bool, err error) {
+	maxRetries := 5
+	baseTime := 2500 // milliseconds
+	backoff := func(i int) {
```
rpk already uses github.com/avast/retry-go i think it would tidier to use it here, its nice and verbose where the below you can miss where you're doing an exponential retry

c-julin replied:
```go
 err = retry.Do(
                func() error {
                        shadowLink, err := cloudClient.ShadowLink.GetShadowLink(ctx, connect.NewRequest(&controlplanev1beta2.GetShadowLinkRequest{
                                Id: linkID,
                        }))
                        if err != nil {
                                zap.L().Sugar().Debugf("unable to get Shadow Link %q, retrying: %v", linkID, err)
                                return err
                        }
                        switch shadowLink.Msg.GetShadowLink().GetState() {
                        case controlplanev1beta2.ShadowLink_STATE_ACTIVE:
                                return nil // Success!
                        case controlplanev1beta2.ShadowLink_STATE_CREATION_FAILED:
                                finalErr = errors.New("an error occurred while creating the shadow link.\nPlease run the command again or contact Redpanda Cloud Support")
                                return retry.Unrecoverable(finalErr) // Don't retry
                        default:
                                return errors.New("shadow link not ready yet") // Retry
                        }
                },
                retry.Attempts(5),
                retry.DelayType(retry.BackOffDelay),
                retry.Delay(2500*time.Millisecond),
                retry.LastErrorOnly(true),
                retry.Context(ctx),
        )

```

c-julin commented on src/go/rpk/pkg/publicapi/controlplane.go:
```diff
@@ -176,6 +192,25 @@ func (cpCl *CloudClientSet) ClusterForID(ctx context.Context, ID string) (*contr
 	return c.Msg.Cluster, nil
 }
 
+func (cpCl *CloudClientSet) ShadowLinkListItemsByDestinationID(ctx context.Context, destID string) ([]*controlplanev1beta2.ShadowLinkListItem, error) {
```
my opinion on this is you should expose the full filter and the caller should just pass destination id

c-julin commented on src/go/rpk/pkg/oauth/authtoken/errors.go:
```diff
@@ -0,0 +1,32 @@
+// Copyright 2024 Redpanda Data, Inc.
```
2025

c-julin committed:
rpk/shadow: add cloud support to failover command

Add dataplane API support to the shadow failover command for cloud
profiles. The implementation follows a hybrid approach:

- Controlplane API is used optionally for GetShadowLink to display
  shadow link details during confirmation prompts
- Dataplane API is used for the actual FailOver operation
- Errors when fetching display details don't block the operation

Changes:
- Add ShadowLink service (v1alpha3) to DataPlaneClientSet
- Update failover command to detect cloud profiles
- Create dataplane client for cloud failover operations
- Add printCloudOverview() to describe.go for cloud shadow link display
- Keep existing admin API path for non-cloud profiles

The failover command now seamlessly supports both cloud and on-prem
deployments.

(cherry picked from commit 138baa0725b058a32df096c517c60b7e446315cf)

r-vasquez committed:
rpk: Add cloud support for shadow delete

r-vasquez committed:
rpk: Add cloud support for shadow describe

It requires a different set of methods
to print the output as the protos from
cloud are different from self-hosted.

r-vasquez commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--3b95f960-2a4a-4cbc-a8da-88f888d03983-->
<!--TEST RESULTS-->
<!--942eb345-af2c-4937-9db5-62287bccdfe6-->
<!--TEST RESULTS-->
<!--c5e80a26-e9a0-409e-b613-97f436619a1a-->
<!--TEST RESULTS-->
<!--703687c9-33a9-4cc5-9ce3-65b4bc9eab81-->
<!--TEST RESULTS-->
<!--c63e115d-f727-406f-8daf-83f6fc142f33-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77095">build#77095</a></summary>


| test_class        | test_method          | test_arguments                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                         |
|:------------------|:---------------------|:----------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------|
| TopicRecoveryTest | test_many_partitions | {"check_mode": "no_check", "cloud_storage_type": 2} | integration | https://buildkite.com/redpanda/redpanda/builds/77095#019ac7e5-7289-4545-9b24-82ada8a6fc5c | FLAKY         | 20/21    | upstream reliability is '99.74226804123711'. current run reliability is '95.23809523809523'. drift is 4.50417 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TopicRecoveryTest&test_method=test_many_partitions |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77376">build#77376</a></summary>


| test_class                    | test_method                      | test_arguments                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                   |
|:------------------------------|:---------------------------------|:--------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_topic_delete                | {"source_cluster_spec": {"cluster_type": "redpanda"}}                           | integration | https://buildkite.com/redpanda/redpanda/builds/77376#019aec94-7c68-41a9-9c5e-da3461e79abc | FLAKY         | 13/21    | upstream reliability is '100.0'. current run reliability is '61.904761904761905'. drift is 38.09524 and the allowed drift is set to 50. The test should PASS           | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_topic_delete  |
| DatalakeDLQTest               | test_dlq_table_for_mixed_records | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1, "query_engine": "spark"} | integration | https://buildkite.com/redpanda/redpanda/builds/77376#019aec9b-49dc-42f5-86e2-c6e03d0774ea | FLAKY         | 19/21    | upstream reliability is '99.70326409495549'. current run reliability is '90.47619047619048'. drift is 9.22707 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeDLQTest&test_method=test_dlq_table_for_mixed_records |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77554">build#77554</a></summary>


| test_class               | test_method                           | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:-------------------------|:--------------------------------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommissioning_rebalancing_node | {"shutdown_decommissioned": true} | integration | https://buildkite.com/redpanda/redpanda/builds/77554#019b009c-ec7c-4005-8480-7f825ff179fd | FLAKY         | 24/31    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1010, p0=0.0760, reject_threshold=0.0100. adj_baseline=0.2734, p1=0.2483, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77597">build#77597</a></summary>


| test_class                 | test_method                           | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                      | test_history                                                                                                                                                                   |
|:---------------------------|:--------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeClusterRestoreTest | test_slow_tiered_storage_dupe_records | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/77597#019b047e-657b-44d4-895a-9164bc2d90db | FLAKY         | 50/51    | Test PASSES after retries.[Inconclusive result after max retries](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#Case-F%3A-inconclusive-result-after-max-retries-(run-passes)%5BinlineExtension%5D)(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0000, p1=1.0000, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_slow_tiered_storage_dupe_records |
| NodesDecommissioningTest   | test_decommissioning_rebalancing_node | {"shutdown_decommissioned": true}                      | integration | https://buildkite.com/redpanda/redpanda/builds/77597#019b0483-b367-4c4b-8ce2-1b7ca9453bd6 | FLAKY         | 13/21    | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0988, p0=0.0022, reject_threshold=0.0100)                                            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77605">build#77605</a></summary>


| test_class                       | test_method                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                              |
|:---------------------------------|:---------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77605#019b0501-853f-427a-8982-cf75eb86ba5e | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0550, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1562, p1=0.1829, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<!--/body-->

r-vasquez committed:
rpk: bump cloud shadow protos to v1

This includes a renaming of `destination_` to
`source_` in the API and in the help text of rpk.

r-vasquez committed:
rpk: add cloud support for shadow status

This includes also support for --format json/yaml

r-vasquez committed:
rpk: cloud support for shadow update

r-vasquez committed:
rpk: specify the need for secret store in shadow create

This will preemptively warn the users if they are
using plain passwords instead of secrets in the
secret store. Also, guiding them through the
relevant rpk command to create one.

r-vasquez committed:
rpk: use pagination when listink shadow topics from cloud

r-vasquez committed:
rpk shadow: add tests - minor cosmetic fixes

r-vasquez committed:
rpk shadow: bump cloud dependencies to v1

this includes dataplane to v1 and the addition of
effective configurations to controlplane response.

Since cloud protos also depends on Core protos, we
had to upgrade proto and it included a new auth
mechanism: PLAIN, which we now support.

r-vasquez committed:
rpk shadow status: bump dataplane

This change includes now the per-partition response
in the ListShadowLinkTopic.

Also, we are now sorting both responses by topic
name and partition ID.

r-vasquez committed:
rpk shadow create: prevent empty rgID in UI

r-vasquez committed:
rpk: sl delete warn for cloud clusters

Cloud clusters use force deletion by default,
so we print an additional warning on deletion

r-vasquez commented on src/go/rpk/pkg/cli/shadow/shadow.go:
```diff
@@ -43,3 +56,39 @@ cluster metadata.
+// Shadow Link operation. (e.g., create, delete, update).
+func waitForOperation(ctx context.Context, cloudClient *publicapi.CloudClientSet, opID string) (isCompleted bool, err error) {
+	backoff := func(i int) {
+		sleepTime := retryDelay * (1 << i) // Exponential backoff.
```
We only do 5 retries. I will cap at 30s

r-vasquez committed:
rpk shadow: misc improvements.

- Cap retry delay for operations to 30s. These ops
  tend to be ready after 20s most of the time.
- Abstract maxPages for paginated requests to a
  package constant.
- Clarify mapper documentation for cloud TLS
  settings.

r-vasquez committed:
rpk: bump Public API deps

This include 2 major changes for rpk shadow:

- Resource Group ID is no longer required.
- Partition info in Cloud now matches the same
  type as in Core.

c-julin commented on src/go/rpk/pkg/cli/shadow/config.go:
```diff
@@ -43,16 +43,20 @@ func newShadowConfigCommand(fs afero.Fs, p *config.Params) *cobra.Command {
 
 By default, this command creates a sample configuration file with placeholder
-values that you customize for your environment.
+values that you customize for your environment. If you are generating a Shadow
```
nit: you can customize

c-julin commented on src/go/rpk/pkg/cli/shadow/config.go:
```diff
@@ -111,12 +114,13 @@ Save the template with documentation to a file:
-func generateSampleConfig() *ShadowLinkConfig {
-	return &ShadowLinkConfig{
+func generateSampleConfig(cloud bool) *ShadowLinkConfig {
+	slcfg := &ShadowLinkConfig{
```
go convention would be shadowLinkCfg but in any case i still think slCfg at least to make the var naming clearer

c-julin commented on src/go/rpk/pkg/cli/shadow/config.go:
```diff
@@ -196,6 +200,29 @@ func generateSampleConfig() *ShadowLinkConfig {
+		// This is confusing on Cloud, user is already providing the Redpanda ID.
+		slcfg.ClientOptions.SourceClusterID = ""
+		// Cloud only accepts passwords from the secret store.
+		slcfg.ClientOptions.AuthenticationConfiguration.ScramConfiguration.Password = "${secrets.PASSWORD_FROM_SECRET_STORE}"
```
I think due to recent confusions should we make this PASSWORD FROM SHADOW SECRET STORE

simon0191 commented on src/go/rpk/pkg/cli/shadow/config.go:
```diff
@@ -196,6 +200,27 @@ func generateSampleConfig() *ShadowLinkConfig {
+			Enabled: true,
+			TLSPEMSettings: &TLSPEMSettings{
+				CA:   "-----BEGIN CERTIFICATE-----\n...\n-----END CERTIFICATE-----",
+				Key:  "-----BEGIN PRIVATE KEY-----\n...\n-----END PRIVATE KEY-----",
```
This should reference a dataplane secret

r-vasquez replied:
The 3 elements? Or just the key?

r-vasquez replied:
For future reference: just the key.

simon0191 replied:
Just the key

c-julin commented on src/go/rpk/pkg/cli/shadow/create.go:
```diff
@@ -141,8 +187,39 @@ func validateParsedShadowLinkConfig(slCfg *ShadowLinkConfig) error {
+			return errors.New("only one of start_at_latest, start_at_earliest, or start_at_timestamp can be provided")
+		}
+	}
+	if slc := slCfg.CloudOptions; slc != nil {
```
small nit: you can technically return here and line 200 early instead of the nested ifs

c-julin commented on src/go/rpk/pkg/cli/shadow/delete.go:
```diff
@@ -55,33 +60,68 @@ Force delete a Shadow Link with active shadow topics:
+				if !noConfirm && !forceDelete {
+					msg := "Are you sure you want to delete this Shadow Link?"
+					if isCloud {
+						msg = "Are you sure you want to delete this Shadow Link? This action is not recoverable and will fail over all the synced shadow topics."
```
is this statement correct? technically we're not failing over its just deleting the shadowlink, just worth double checking with docs 
cc: @kbatuigas

r-vasquez replied:
IIUC to @simon0191 by default we are force deleting (which failovers + delete the link)

c-julin commented on src/go/rpk/pkg/cli/shadow/describe.go:
```diff
@@ -39,6 +52,10 @@ or all sections of the configuration.
 consumer offset synchronization, and security synchronization settings.
+
+For Redpanda Cloud, rpk will use the Redpanda ID of the cluster you are already
+logged on. If you wish to use a different one either login and create a profile
```
currently logged into.

i think reads better instead of already logged on

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/77597-->
<!--d05d2e54-d920-45cf-a0a6-0cf907365bc4-->
<!--title-->
## Retry command for [Build#77597](https://buildkite.com/redpanda/redpanda/builds/77597)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_decommissioning_rebalancing_node@{"shutdown_decommissioned":true}
```
<!--/body-->

vbotbuildovich commented:
/backport v25.3.x
